python/singa/autograd.py

```diff
@@ -2760,3 +2760,107 @@ def backward(self, dy):
 
 def reciprocal(x):
     return Reciprocal()(x)[0]
+
+
+class Gemm(Operation):
+    def __init__(self, alpha=1.0, beta=1.0, transA=0, transB=0):
+        """
+        init a General Matrix multiplication(Gemm) operator
+        Compute Y = alpha * A' * B' + beta * C, where input tensor A has shape (M, K) or (K, M), input tensor B has shape (K, N) or (N, K), input tensor C is broadcastable to shape (M, N), and output tensor Y has shape (M, N).
+        A' = transpose(A) if transA else A
+        B' = transpose(B) if transB else B
+        Args:alpha: 
+            float, Scalar multiplier for the product of input tensors A * B.
+        Args:beta: 
+            float, Scalar multiplier for input tensor C.
+        Args:transA: 
+            int, Whether A should be transposed
+        Args:transB: 
+            int, Whether B should be transposed
+        Returns: 
+            tensor, the output
+        """
+        super(Gemm, self).__init__()
+        self.alpha = alpha
+        self.beta = beta
+        self.transA = transA
+        self.transB = transB
+
+    def forward(self, A, B, C=None):
+        """
+        forward propogation of Gemm
+        Args:A: 
+            tensor, The shape of A should be (M, K) if transA is 0, or (K, M) if transA is non-zero.
+        Args:B: 
+            tensor, The shape of B should be (K, N) if transB is 0, or (N, K) if transB is non-zero.
+        Args:C: 
+            tensor(optional), Optional input tensor C. If not specified, the computation is done as if C is a scalar 0. The shape of C should be unidirectional broadcastable to (M, N).
+        Returns: 
+            tensor, the output
+        """
+        _A = singa.DefaultTranspose(A) if self.transA == 1 else A
+        _B = singa.DefaultTranspose(B) if self.transB == 1 else B
+        if training:
+            self.inputs = (_A, _B, C)
+        tmpM = singa.MultFloat(singa.Mult(_A, _B), self.alpha)
```

Comment:
Did you try [MultiWithScale](https://github.com/apache/singa/blob/master/src/api/core_tensor.i#L319)?
Or the implementation in this [PR](https://github.com/apache/singa/pull/515/files#diff-a236c495700ad62b39e28a4ece74711fR2318)

Reply:
The MultiWithScale needs 5 params(alpha, A, B, beta, C), if the user doesn't input the C (C is None), I guess I still need to init a C with zero values...

Reply:
MultiWithScale needs 5 inputs, (alpha, A, B, beta, C), if the user doesn't input a C, which means C is none, it still needs to init a C with all zero values. maybe we should optimize the MultiWithScale?

Reply:
how to optimize it?
I think MultiWithScale is more efficient as it calls the cpp/cuda gemm directly instead of doing the matrix multiplication and addition separately..

Reply:
I just try this MultiWithScale, but it cannot support the Unidirectional Broadcasting for bias. https://github.com/onnx/onnx/blob/master/docs/Broadcasting.md

Reply:
and also, please check the test cases for gemm, https://github.com/onnx/onnx/blob/master/onnx/backend/test/case/node/gemm.py#L27-L108

Reply:
I see. This PR can pass the tests?
Pls fix the conflicts and format the code.

Reply:
Yes, since shicong has re-implement the add and mul of singa to support it.
got it.